src/primary_names.lua

```diff
@@ -50,8 +50,60 @@ NextPrimaryNamesPruneTimestamp = NextPrimaryNamesPruneTimestamp or 0
 --- @field fundingPlan table
 --- @field fundingResult table
 
-local function baseNameForName(name)
-	return (name or ""):match("[^_]+$") or name
+-- NOTE: lua 5.3 has limited regex support, particularly for lookaheads and negative lookaheads or use of {n}
+---@param name string
+---@description Asserts that the provided name is a valid undername
+---@example
+---```lua
+---utils.assertValidateUndername("my-undername")
+---```
+function primaryNames.assertValidUndername(name)
+	--- RULES FOR UNDERNAMES
+	--- min 1 char
+	--- max 61 chars
+	--- no starting dashes or underscores
+	--- alphanumeric, dashes, underscores OR one '@' sign
+
+	local validLength = #name <= constants.MAX_UNDERNAME_LENGTH
+	assert(validLength, "Undername is too long, recieved length of " .. tostring(#name))
+	local validRegex = string.match(name, "^[a-zA-Z0-9]*$") ~= nil
+		or string.match(name, "^[a-zA-Z0-9]+[a-zA-Z0-9_-]*$") ~= nil
+	local valid = validLength and validRegex
+	assert(valid, "Invalid undername " .. name)
+end
+
+--- Asserts that a name is a valid Primary name
+--- Validates the undername and base name
+--- @param name string The name to check
+function primaryNames.assertValidPrimaryName(name)
+	assert(name and type(name) == "string", "Name is required and must be a string.")
+
+	assert(
+		#name <= constants.MAX_PRIMARY_NAME_LENGTH,
+		"Primary Name is too long, received length of " .. tostring(#name)
+	)
+
+	-- Ensure name starts and ends with alphanumeric, and only allows `_` or `-` in between
+	assert(
+		string.match(name, "^[a-zA-Z0-9]$") ~= nil
```

Comment:
put this regex into constants file

Reply:
and please confirm with GPT that lua supports these unioned regexes - there may be a more concise regex available